### PR TITLE
Fix iter_payday_events test

### PR DESCRIPTION
**Pull request by whit537**
Reticketing from https://github.com/gratipay/gratipay.com/pull/4232#issuecomment-267676317.

### CODE REVIEW

**Comment by whit537** [comment]
```
gratipay-test=# select * from paydays;                                                                        ┌────┬───────────────────────────────┬───────────────────────────────┬────────┬────────┬───────┬────────┐
│ id │           ts_start            │            ts_end             │ volume │ nusers │ stage │ nteams │
├────┼───────────────────────────────┼───────────────────────────────┼────────┼────────┼───────┼────────┤
│ 25 │ 2016-12-16 20:39:15.226107+00 │ 2016-12-16 20:39:15.538988+00 │   0.00 │      0 │     2 │      0 │
│ 26 │ 2016-12-16 20:39:15.581323+00 │ 2016-12-16 20:39:16.004417+00 │  10.00 │      2 │     2 │      1 │
│ 27 │ 2016-12-23 20:39:16.022322+00 │ 2016-12-23 20:39:16.455521+00 │  10.00 │      2 │     2 │      1 │
│ 28 │ 2016-12-30 20:39:16.47454+00  │ 1970-01-15 00:00:00+00        │   0.00 │      0 │     0 │      0 │
└────┴───────────────────────────────┴───────────────────────────────┴────────┴────────┴───────┴────────┘
(4 rows)

gratipay-test=#
```

```diff
diff --git a/tests/py/test_history.py b/tests/py/test_history.py
index 178d0d3..a9771e3 100644
--- a/tests/py/test_history.py
+++ b/tests/py/test_history.py
@@ -90,9 +90,10 @@ class TestHistory(BillingHarness):
 
         events = list(iter_payday_events(self.db, obama))
         assert events[0]['given'] == 20
+        import pdb; pdb.set_trace()
         assert len(events) == 11
 
-    def test_iter_payday_events_with_failed_exchanges(self):
+    def est_iter_payday_events_with_failed_exchanges(self):
         alice = self.make_participant('alice', claimed_time='now')
         self.make_exchange('braintree-cc', 50, 0, alice)
         self.make_exchange('braintree-cc', 12, 0, alice, status='failed')

```

**Comment by whit537** [comment]
```
gratipay-test=# select * from payments;
┌────┬───────────────────────────────┬─────────────┬───────────────┬────────┬────────────────┬────────┐
│ id │           timestamp           │ participant │     team      │ amount │   direction    │ payday │
├────┼───────────────────────────────┼─────────────┼───────────────┼────────┼────────────────┼────────┤
│ 33 │ 2016-12-16 20:47:24.009245+00 │ obama       │ TheEnterprise │  10.00 │ to-team        │     34 │
│ 34 │ 2016-12-16 20:47:24.009245+00 │ picard      │ TheEnterprise │  10.00 │ to-participant │     34 │
│ 35 │ 2016-12-23 20:47:24.459356+00 │ obama       │ TheEnterprise │  10.00 │ to-team        │     35 │
│ 36 │ 2016-12-23 20:47:24.459356+00 │ picard      │ TheEnterprise │  10.00 │ to-participant │     35 │
└────┴───────────────────────────────┴─────────────┴───────────────┴────────┴────────────────┴────────┘
(4 rows)

gratipay-test=# select id, timestamp from exchanges;
┌────┬───────────────────────────────┐
│ id │           timestamp           │
├────┼───────────────────────────────┤
│  1 │ 2016-12-16 20:47:24.174547+00 │
│  2 │ 2016-12-16 20:47:24.623078+00 │
└────┴───────────────────────────────┘
(2 rows)

gratipay-test=# select * from transfers;
┌────┬───────────┬────────┬────────┬────────┬─────────┬────────┐
│ id │ timestamp │ tipper │ tippee │ amount │ context │ payday │
├────┼───────────┼────────┼────────┼────────┼─────────┼────────┤
└────┴───────────┴────────┴────────┴────────┴─────────┴────────┘
(0 rows)

gratipay-test=#
```

**Comment by whit537** [comment]
Interesting. I set my clock back to find a point of comparison, and I'm still seeing December dates for _some_ of the data. I guess that's where we depend on `now` in Postgres, which I haven't restarted since changing the date?

```
gratipay-test=# select * from paydays;
┌────┬───────────────────────────────┬───────────────────────────────┬────────┬────────┬───────┬────────┐
│ id │           ts_start            │            ts_end             │ volume │ nusers │ stage │ nteams │
├────┼───────────────────────────────┼───────────────────────────────┼────────┼────────┼───────┼────────┤
│ 47 │ 2016-12-23 20:54:08.168849+00 │ 2016-12-23 20:54:08.612561+00 │  10.00 │      2 │     2 │      1 │
│ 45 │ 2016-12-16 20:54:07.350908+00 │ 2016-12-16 20:54:07.669071+00 │   0.00 │      0 │     2 │      0 │
│ 46 │ 2016-12-16 20:54:07.715133+00 │ 2016-12-16 20:54:08.14544+00  │  10.00 │      2 │     2 │      1 │
│ 48 │ 2016-12-30 20:54:08.633357+00 │ 1970-02-12 00:00:00+00        │   0.00 │      0 │     0 │      0 │
└────┴───────────────────────────────┴───────────────────────────────┴────────┴────────┴───────┴────────┘
(4 rows)

gratipay-test=# select * from payments;
┌────┬───────────────────────────────┬─────────────┬───────────────┬────────┬────────────────┬────────┐
│ id │           timestamp           │ participant │     team      │ amount │   direction    │ payday │
├────┼───────────────────────────────┼─────────────┼───────────────┼────────┼────────────────┼────────┤
│ 45 │ 2016-12-16 20:54:07.722936+00 │ obama       │ TheEnterprise │  10.00 │ to-team        │     46 │
│ 46 │ 2016-12-16 20:54:07.722936+00 │ picard      │ TheEnterprise │  10.00 │ to-participant │     46 │
│ 47 │ 2016-12-23 20:54:08.186376+00 │ obama       │ TheEnterprise │  10.00 │ to-team        │     47 │
│ 48 │ 2016-12-23 20:54:08.186376+00 │ picard      │ TheEnterprise │  10.00 │ to-participant │     47 │
└────┴───────────────────────────────┴─────────────┴───────────────┴────────┴────────────────┴────────┘
(4 rows)

gratipay-test=# select id, timestamp from exchanges;
┌────┬───────────────────────────────┐
│ id │           timestamp           │
├────┼───────────────────────────────┤
│  1 │ 2016-11-18 20:54:07.887496+00 │
│  2 │ 2016-11-18 20:54:08.360015+00 │
└────┴───────────────────────────────┘
(2 rows)

gratipay-test=# select * from transfers;
┌────┬───────────┬────────┬────────┬────────┬─────────┬────────┐
│ id │ timestamp │ tipper │ tippee │ amount │ context │ payday │
├────┼───────────┼────────┼────────┼────────┼─────────┼────────┤
└────┴───────────┴────────┴────────┴────────┴─────────┴────────┘
(0 rows)

gratipay-test=#
```

**Comment by whit537** [comment]
I feel like we did this last year.

**Comment by whit537** [comment]
> I'm still seeing December dates for _some_ of the data.

Yet the test still passes!

**Comment by whit537** [comment]
# Working

```
[{'given': Decimal('20.00'), 'kind': 'totals', 'received': 0},
 {'balance': Decimal('0.00'),
  'date': datetime.date(2016, 12, 23),
  'kind': 'day-open',
  'payday_number': 2},
 {'amount': Decimal('10.00'),
  'balance': Decimal('0.00'),
  'direction': 'to-team',
  'id': 51L,
  'kind': 'payment',
  'participant': u'obama',
  'payday': 51,
  'team': u'TheEnterprise',
  'timestamp': datetime.datetime(2016, 12, 23, 20, 58, 30, 938757, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('10.00'), 'kind': 'day-close'},
 {'balance': Decimal('10.00'),
  'date': datetime.date(2016, 12, 16),
  'kind': 'day-open',
  'payday_number': 1},
 {'amount': Decimal('10.00'),
  'balance': Decimal('10.00'),
  'direction': 'to-team',
  'id': 49L,
  'kind': 'payment',
  'participant': u'obama',
  'payday': 50,
  'team': u'TheEnterprise',
  'timestamp': datetime.datetime(2016, 12, 16, 20, 58, 30, 482858, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('20.00'), 'kind': 'day-close'},
 {'balance': Decimal('20.00'),
  'date': datetime.date(2016, 11, 18),
  'kind': 'day-open'},
 {'amount': Decimal('10.00'),
  'balance': Decimal('20.00'),
  'fee': Decimal('0.61'),
  'id': 2,
  'kind': 'charge',
  'note': None,
  'participant': u'obama',
  'recorder': None,
  'ref': None,
  'route': 57L,
  'status': 'succeeded',
  'timestamp': datetime.datetime(2016, 11, 18, 20, 58, 31, 116590, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'),
  'balance': Decimal('10.00'),
  'fee': Decimal('0.61'),
  'id': 1,
  'kind': 'charge',
  'note': None,
  'participant': u'obama',
  'recorder': None,
  'ref': None,
  'route': 57L,
  'status': 'succeeded',
  'timestamp': datetime.datetime(2016, 11, 18, 20, 58, 30, 651852, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('0.00'), 'kind': 'day-close'}]
```

# Borked

```
[{'given': Decimal('20.00'), 'kind': 'totals', 'received': 0},
 {'balance': Decimal('0.00'),
  'date': datetime.date(2016, 12, 23),
  'kind': 'day-open',
  'payday_number': 2},
 {'amount': Decimal('10.00'),
  'balance': Decimal('0.00'),
  'direction': 'to-team',
  'id': 55L,
  'kind': 'payment',
  'participant': u'obama',
  'payday': 55,
  'team': u'TheEnterprise',
  'timestamp': datetime.datetime(2016, 12, 23, 21, 0, 17, 386676, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('10.00'), 'kind': 'day-close'},
 {'balance': Decimal('10.00'),
  'date': datetime.date(2016, 12, 16),
  'kind': 'day-open',
  'payday_number': 1},
 {'amount': Decimal('10.00'),
  'balance': Decimal('10.00'),
  'fee': Decimal('0.61'),
  'id': 2,
  'kind': 'charge',
  'note': None,
  'participant': u'obama',
  'recorder': None,
  'ref': None,
  'route': 60L,
  'status': 'succeeded',
  'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 17, 551729, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'),
  'balance': Decimal('0.00'),
  'fee': Decimal('0.61'),
  'id': 1,
  'kind': 'charge',
  'note': None,
  'participant': u'obama',
  'recorder': None,
  'ref': None,
  'route': 60L,
  'status': 'succeeded',
  'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 16, 576675, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'),
  'balance': Decimal('-10.00'),
  'direction': 'to-team',
  'id': 53L,
  'kind': 'payment',
  'participant': u'obama',
  'payday': 54,
  'team': u'TheEnterprise',
  'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 16, 357140, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('0.00'), 'kind': 'day-close'}]
```

**Comment by whit537** [comment]
# Working

```
[{'given': Decimal('20.00'), 'kind': 'totals', 'received': 0},
 {'balance': Decimal('0.00'), 'date': datetime.date(2016, 12, 23), 'kind': 'day-open', 'payday_number': 2},
 {'amount': Decimal('10.00'), 'balance': Decimal('0.00'), 'direction': 'to-team', 'id': 51L, 'kind': 'payment', 'participant': u'obama', 'payday': 51, 'team': u'TheEnterprise', 'timestamp': datetime.datetime(2016, 12, 23, 20, 58, 30, 938757, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('10.00'), 'kind': 'day-close'},
 {'balance': Decimal('10.00'), 'date': datetime.date(2016, 12, 16), 'kind': 'day-open', 'payday_number': 1},
 {'amount': Decimal('10.00'), 'balance': Decimal('10.00'), 'direction': 'to-team', 'id': 49L, 'kind': 'payment', 'participant': u'obama', 'payday': 50, 'team': u'TheEnterprise', 'timestamp': datetime.datetime(2016, 12, 16, 20, 58, 30, 482858, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('20.00'), 'kind': 'day-close'},
 {'balance': Decimal('20.00'), 'date': datetime.date(2016, 11, 18), 'kind': 'day-open'},
 {'amount': Decimal('10.00'), 'balance': Decimal('20.00'), 'fee': Decimal('0.61'), 'id': 2, 'kind': 'charge', 'note': None, 'participant': u'obama', 'recorder': None, 'ref': None, 'route': 57L, 'status': 'succeeded', 'timestamp': datetime.datetime(2016, 11, 18, 20, 58, 31, 116590, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'), 'balance': Decimal('10.00'), 'fee': Decimal('0.61'), 'id': 1, 'kind': 'charge', 'note': None, 'participant': u'obama', 'recorder': None, 'ref': None, 'route': 57L, 'status': 'succeeded', 'timestamp': datetime.datetime(2016, 11, 18, 20, 58, 30, 651852, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('0.00'), 'kind': 'day-close'}]
```

# Borked

```
[{'given': Decimal('20.00'), 'kind': 'totals', 'received': 0},
 {'balance': Decimal('0.00'), 'date': datetime.date(2016, 12, 23), 'kind': 'day-open', 'payday_number': 2},
 {'amount': Decimal('10.00'), 'balance': Decimal('0.00'), 'direction': 'to-team', 'id': 55L, 'kind': 'payment', 'participant': u'obama', 'payday': 55, 'team': u'TheEnterprise', 'timestamp': datetime.datetime(2016, 12, 23, 21, 0, 17, 386676, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('10.00'), 'kind': 'day-close'},
 {'balance': Decimal('10.00'), 'date': datetime.date(2016, 12, 16), 'kind': 'day-open', 'payday_number': 1},
 {'amount': Decimal('10.00'), 'balance': Decimal('10.00'), 'fee': Decimal('0.61'), 'id': 2, 'kind': 'charge', 'note': None, 'participant': u'obama', 'recorder': None, 'ref': None, 'route': 60L, 'status': 'succeeded', 'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 17, 551729, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'), 'balance': Decimal('0.00'), 'fee': Decimal('0.61'), 'id': 1, 'kind': 'charge', 'note': None, 'participant': u'obama', 'recorder': None, 'ref': None, 'route': 60L, 'status': 'succeeded', 'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 16, 576675, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'amount': Decimal('10.00'), 'balance': Decimal('-10.00'), 'direction': 'to-team', 'id': 53L, 'kind': 'payment', 'participant': u'obama', 'payday': 54, 'team': u'TheEnterprise', 'timestamp': datetime.datetime(2016, 12, 16, 21, 0, 16, 357140, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=0, name=None))},
 {'balance': Decimal('0.00'), 'kind': 'day-close'}]
```

**Comment by whit537** [comment]
# Working

```
[{'balance': Decimal('20.00'), 'kind': 'day-close'},
 {'balance': Decimal('20.00'), 'date': datetime.date(2016, 11, 18), 'kind': 'day-open'}]
```

# Borked

```
[]
```

**Comment by whit537** [comment]
Alright, so the working version has an extra day boundary? While going through that I noticed that one of the `amount`s seemed out of order?

**Comment by whit537** [comment]
I'm close on this one. It's a weird interaction that may actually only show up on December 16. 😳

Notes from the bus:

```
There are a few things going on:
    ts_end shouldn't be incremented if it's the epoch
    we're not updating exchanges
    wtf is the algorithm here?
    is the xxxx-12-16 date significant somehow?
    we're not crossing a day boundary, is that it?

    maybe factor out a shift_payday function to update times
```

**Comment by whit537** [comment]
Alright, ready to go here.

**Comment by whit537** [comment]
Rebased, was 563a03b02013c473f23c44521b833aef4c00345e.

**Comment by whit537** [comment]
@kaguillera Any chance you could review this one?

**Comment by whit537** [comment]
Rebased, was b87633e7255c3aa40a89440805c486850928d001.

**Comment by whit537** [comment]
I am working on landing #4232. I had made this the base for that, so in order to get that to master I am going to go ahead and land this as well.

**Comment by whit537** [comment]
Actually, I'm not quite sure of the state of this branch. It appears to have an old version of #4323 incorporated? Investigating ...

File: tests/py/test_history.py [hunk 1]
```diff
@@ -66,16 +73,13 @@ def test_iter_payday_events(self):
...
-            UPDATE transfers
-                SET timestamp = "timestamp" + interval %(delta)s;
-        """, dict(delta=delta))
+        if now.month < 2:
+
+            # Above, we created four paydays in sequential weeks, with the
+            # latest being today. We don't want them to go back into the
+            # previous year, so let's shift them forward if it's close.
+
+            self.shift_all_paydays_by('3 months')
```
Comment (kaguillera) [inline tests/py/test_history.py]: If now.month is 1 (i.e. January) isn't `shift_all_paydays_by` going to push the `ts_start` and `ts_end` into the future? Is this the effect that you want? Unless this is just initializing the beginning/first date for the set of fake paydays that you intend to produce?

Comment (kaguillera) [inline tests/py/test_history.py]: Answered IRL

**Comment by whit537** [comment]
Okay, it's just that the three commits from this PR are interleaved with the commits from #4232.

**Comment by whit537** [comment]
I've rebased here with a squashed #4232. Previous was 33f8ba70787b76198e03f90d81e103d7342a3240.

File: tests/py/test_history.py (hunk 1 above)
Comment (whit537) [inline tests/py/test_history.py]: Yes, that's the effect. I think for testing purposes it doesn't matter so much whether it's in the future or the past.

**Comment by whit537** [comment]
And rebased on to latest `master` ...

**Comment by whit537** [comment]
@kaguillera You ready to merge this when green?

**Comment by whit537** [comment]
💃

**Pull request merged**